oauth2client/dictionary_storage.py

```diff
@@ -0,0 +1,78 @@
+# Copyright 2014 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Dictionary storage for OAuth2 Credentials."""
+
+from oauth2client.client import OAuth2Credentials
+from oauth2client.locked_storage import LockedStorage
+
+
+__author__ = 'jonwayne@google.com (Jon Wayne Parrott)'
+
+
+class DictionaryStorage(LockedStorage):
+    """Store and retrieve credentials to and from a dictionary-like object."""
+
+    def __init__(self, dictionary, key, lock=None):
+        """Construct a DictionaryStorage instance.
+
+        Args:
+            dictionary: A dictionary or dictionary-like object.
+            key: A hashable or a function returning a hashable. The credentials
```

Comment:
I'm okay with removing this here, but the intention is to facilitate creating storage classes where the key is contextual, for example:

```python
storage = DictionaryStorage(session, key=lambda: users.get_current_user().email)
```

However, this contextualization is typically experienced when using the web framework helpers (like `flask_util`) and this can be pushed into that layer, like so:

```python

oauth2 = UserOAuth2(storage=lambda request: DictionaryStorage(flask.session, key=request.user_id))
```

Either way, we must provide some mechanism for this.